Simplify the UI of the `HilbertSchmidt` and `LocalHilbertSchmidt` templates (#7933)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>Breaking changes 💔</h3>

* The `qml.HilbertSchmidt` and `qml.LocalHilbertSchmidt` templates have been updated and their UI has been remarkably simplified. 

>   They now accept a list of operations instead of a `qml.tape.QuantumScript` of operations and quantum functions with separate parameters and wires.

Reviewer: ```suggestion
  They now accept an operation or a list of operations instead of a `qml.tape.QuantumScript` of operations and quantum functions with separate parameters and wires.
```

  Instead, the wires and parameters of the provided unitaries are inferred from the inputs.

  As a concrete example, this is how the Hilbert-Schmidt Test cost between the unitary `U` and an approximate unitary ``V`` can be evaluated:

  ```python
  U = qml.Hadamard(0)
  V = qml.RZ(0, wires=1)

  dev = qml.device("default.qubit", wires=2)

  @qml.qnode(dev)
  def hilbert_test(V, U):
      qml.HilbertSchmidt(V, U)
      return qml.probs()

  def cost_hst(V, U):

>       return 1 - hilbert_test(V, U)[0]

Author: The cost function is the same one defined in the paper (see eq. (5) of https://arxiv.org/pdf/1807.00800), so I would say it makes sense to keep it. What do you think?

>       return 1 - hilbert_test(V, U)[0]

Reviewer: We should probably focus on the "new call signature" part of this breaking change, and not "how to use in it a to do stuff". That is the class docstring.

>       return 1 - hilbert_test(V, U)[0]

Author: I agree, I think this would be more useful for a changelog. I'll update it :+1:

>       return 1 - hilbert_test(V, U)[0]

Author: Done!

<h3>Breaking changes 💔</h3>

* The `qml.HilbertSchmidt` and `qml.LocalHilbertSchmidt` templates have been updated and their UI has been remarkably simplified. 
  They now accept an operation or a list of operations as quantum unitaries.

  Before, the signature of these templates required providing the `U` and `V` unitaries as a `qml.tape.QuantumTape` and a quantum function,
  respectively, along with separate parameters and wires.

  ```python
  with qml.QueuingManager.stop_recording():
      u_tape = qml.tape.QuantumTape([qml.Hadamard(0)])

  def v_function(params):
      qml.RZ(params, wires=1)  

  v_params = 0.1
  v_wires = [1]    
  ```

  ```pycon
  >>> qml.HilbertSchmidt(v_params, v_function=v_function, v_wires=v_wires, u_tape=u_tape)
  HilbertSchmidt(0.1, wires=[0, 1])
  ```

  Instead, now the templates accept one or more operators as `U` and `V` unitaries. 
  The wires and parameters of the approximate unitary `V` are inferred from the inputs, according to the order provided.

>   ```python
>   U = qml.Hadamard(0)
>   V = qml.RZ(0.1, wires=1)
>   ```
> 
>   ```pycon
>   >>> qml.HilbertSchmidt(V, U)
>   HilbertSchmidt(0.1, wires=[0, 1])
>   ```

Reviewer: ```suggestion
  ```python
  >>> U = qml.Hadamard(0)
  >>> V = qml.RZ(0.1, wires=1)
  >>> qml.HilbertSchmidt(V, U)
  HilbertSchmidt(0.1, wires=[0, 1])
  ```
```
A little more concise to just use one block.

<h3>Breaking changes 💔</h3>

> * The `qml.HilbertSchmidt` and `qml.LocalHilbertSchmidt` templates have been updated and their UI has been remarkably simplified. 
>   They now accept an operation or a list of operations as quantum unitaries.
> 
>   Before, the signature of these templates required providing the `U` and `V` unitaries as a `qml.tape.QuantumTape` and a quantum function,
>   respectively, along with separate parameters and wires.
> 
>   ```python
>   with qml.QueuingManager.stop_recording():
>       u_tape = qml.tape.QuantumTape([qml.Hadamard(0)])
> 
>   def v_function(params):
>       qml.RZ(params, wires=1)  
> 
>   v_params = 0.1
>   v_wires = [1]    
>   ```
> 
>   ```pycon
>   >>> qml.HilbertSchmidt(v_params, v_function=v_function, v_wires=v_wires, u_tape=u_tape)
>   HilbertSchmidt(0.1, wires=[0, 1])
>   ```
> 
>   The templates have been improved to accept one or more operators as `U` and `V` unitaries. 
>   The wires and parameters of the approximate unitary `V` are inferred from the inputs, according to the order provided.
> 
>   ```python
>   >>> U = qml.Hadamard(0)
>   >>> V = qml.RZ(0.1, wires=1)
>   >>> qml.HilbertSchmidt(V, U)
>   HilbertSchmidt(0.1, wires=[0, 1])
>   ```
>   [(#7933)](https://github.com/PennyLaneAI/pennylane/pull/7933)

Reviewer 2: ````suggestion
* The `qml.HilbertSchmidt` and `qml.LocalHilbertSchmidt` templates have been updated and their UI has been remarkably simplified. 
  They now accept an operation or a list of operations as quantum unitaries.
  [(#7933)](https://github.com/PennyLaneAI/pennylane/pull/7933)

  In past versions of PennyLane, these templates required providing the `U` and `V` unitaries as a `qml.tape.QuantumTape` and a quantum function,
  respectively, along with separate parameters and wires.

  With this release, each template has been improved to accept one or more operators as  unitaries. 
  The wires and parameters of the approximate unitary `V` are inferred from the inputs, according to the order provided.

  ```python
  >>> U = qml.Hadamard(0)
  >>> V = qml.RZ(0.1, wires=1)
  >>> qml.HilbertSchmidt(V, U)
  HilbertSchmidt(0.1, wires=[0, 1])
  ```
````
